Preact or Svelte user feedback widget blog post (#143)
getsentry/sentry.engineering: data/blog/preact-or-svelte-an-embedded-widget-use-case.md

_Vite bundle visualizer for the Preact build — almost half the build is just the framework_

![Bundle size visualizer for Svelte build](/images/preact-or-svelte-an-embedded-widget-use-case/svelte-build-analyzer.png)

_Vite bundle visualizer for the Svelte build — majority of the build is the feedback dialog and form_

After looking at the builds more closely with the Vite bundle visualizer tool, I noticed that nearly half of the Preact implementation's build size was attributed to the framework itself. Considering our future plans to incorporate screenshots and annotation capabilities, which would significantly increase the build size, the marginal difference in framework size would become negligible, so the slightly smaller build size with Svelte would not help with the final bundle size. Additionally, Preact closely resembles React, which aligns well with our developers' familiarity with React and would greatly improve code maintainability. As a result, we chose Preact as the framework for expanding our User Feedback product to include screenshot and annotation functionalities.

## The Aftermath

> In the end, adding Preact and screenshotting functionality to our user feedback widget increased the bundle size of @sentry/browser with the user feedback integration by 22.93%, from 30.81KB to 37.88KB. Although it’s a pretty large bundle size increase, using Preact allowed us to go from writing code like this:

Reviewer: I actually am concerned of calling it out as an increase of bundle size, being so big without taking the chance to talk about treeshaking. And how that increase assumes all code from the package will be used.

If someone installing `@sentry/browser` doesn't use the feedback code, it'll be not include any of that code. And that's unfortunately not very well understood.

here's a blog post about reducing bundle size and talking about treeshaking you could link to, if you decide to add a note in that area: https://blog.sentry.io/sentry-bundle-size-how-we-reduced-replay-sdk-by-35/

```suggestion
In the end, adding Preact and screenshotting functionality to our user feedback widget increased the bundle size of `[@sentry/browser](https://www.npmjs.com/package/@sentry/browser)` with the user feedback integration by 22.93%, from 30.81KB to 37.88KB. Although it’s a pretty large bundle size increase, using Preact allowed us to go from writing code like this:
```

```js
const emailEl = createElement('input', {
  id: 'email',
  type: showEmail ? 'text' : 'hidden',
  ['aria-hidden']: showEmail ? 'false' : 'true',
  name: 'email',
  required: isEmailRequired,
  className: 'form__input',
  placeholder: emailPlaceholder,
  value: defaultEmail,
})

showEmail &&
  createElement(
    'label',
    {
      htmlFor: 'email',
      className: 'form__label',
    },
    [
      createElement(
        'span',
        { className: 'form__label__text' },
        emailLabel,
        isEmailRequired &&
          createElement('span', { className: 'form__label__text--required' }, ' (required)')
      ),
      emailEl,
    ]
  ),
  !showEmail && emailEl
```

to this:

```js
{
  showEmail ? (
    <label for="email" class="form__label">
      <LabelText label={emailLabel} isRequired={isEmailRequired} />
      <input
        class="form__input"
        defaultValue={defaultEmail}
        id="email"
        name="email"
        placeholder={emailPlaceholder}
        required={isEmailRequired}
        type="text"
      ></input>
    </label>
  ) : (
    <input aria-hidden value={defaultEmail} name="email" type="hidden" />
  )
}
```

> With our end goal of [adding screenshotting and cropping capabilities](https://sentry.io/changelog/user-feedback-widget-screenshots/) to our user feedback widget, using a framework was a necessity. With the use of Preact, implementing screenshotting and cropping became much easier and much more readable. One of our interns was even able to make a surprise contribution and improve on the cropping capabilities. Seeing how our feedback widget turned out in the end, I think we made a great choice in framework.

Reviewer: ```suggestion
With our end goal of [adding screenshotting and cropping capabilities](https://sentry.io/changelog/user-feedback-widget-screenshots/) to our user feedback widget, using a framework was a necessity. With the use of Preact, implementing screenshotting and cropping became much easier and much more readable. One of our interns was even able to make a [surprise contribution and improve on the cropping capabilities](https://github.com/getsentry/sentry-javascript/pull/13071). Seeing how our feedback widget turned out in the end, I think we made a great choice in framework.
```

---
title: 'Preact or Svelte? An Embedded Widget Use Case'
date: '2024-09-06'
tags: ['javascript', 'sdk']
draft: false
summary: 'Preact or Svelte, which framework is best for building an embedded user feedback widget?'
images: [/images/preact-or-svelte-an-embedded-widget-use-case/hero.jpg]
layout: PostLayout
authors: ['catherinelee']
---

> Sentry's [user feedback widget](https://docs.sentry.io/product/user-feedback/#user-feedback-widget) allows anyone to submit feedback quickly and easily any time they encounter something that isn’t working as expected. It’s a form with a name, email, and description fields. However, sometimes a description just isn’t enough to describe an issue in detail. Therefore, we decided to add a button that allows the user to take a screenshot of their web page and crop it, allowing the user to better illustrate their issue.

Reviewer: ```suggestion
Sentry's [user feedback widget](https://docs.sentry.io/product/user-feedback/#user-feedback-widget) allows anyone to submit feedback quickly and easily any time they encounter something that isn’t working as expected. It’s a form with a name, email, and description fields. However, sometimes **a picture is worth a thousand words**. Therefore, we decided to add a button that allows the user to take a screenshot of their web page and crop it, allowing the user to better illustrate their issue.
```

_Vite bundle visualizer for the Preact build — almost half the build is just the framework_

![Bundle size visualizer for Svelte build](/images/preact-or-svelte-an-embedded-widget-use-case/svelte-build-analyzer.png)

_Vite bundle visualizer for the Svelte build — majority of the build is the feedback dialog and form_

After looking at the builds more closely with the Vite bundle visualizer tool, I noticed that nearly half of the Preact implementation's build size was attributed to the framework itself. Considering our future plans to incorporate screenshots and annotation capabilities, which would significantly increase the build size, the marginal difference in framework size would become negligible, so the slightly smaller build size with Svelte would not help with the final bundle size. Additionally, Preact closely resembles React, which aligns well with our developers' familiarity with React and would greatly improve code maintainability. As a result, we chose Preact as the framework for expanding our User Feedback product to include screenshot and annotation functionalities.

## The Aftermath

> In the end, adding Preact and screenshotting functionality to our user feedback widget increased the bundle size of `[@sentry/browser](https://www.npmjs.com/package/@sentry/browser)` with the user feedback integration by 22.93%, from 30.81KB to 37.88KB. Although it’s a pretty large bundle size increase, using Preact allowed us to go from writing code like this:

Reviewer: oops, the style I suggested did break the link:

```suggestion
In the end, adding Preact and screenshotting functionality to our user feedback widget increased the bundle size of [`@sentry/browser`](https://www.npmjs.com/package/@sentry/browser)` with the user feedback integration by 22.93%, from 30.81KB to 37.88KB. Although it’s a pretty large bundle size increase, using Preact allowed us to go from writing code like this:
```
